rfcs/20200609-logsigmoid/README.md

```diff
@@ -0,0 +1,58 @@
+# Introducing logsigmoid operation
+
+## Introduction
+
+Request comes from DGP graph lib community https://www.dgl.ai. Logsigmoid is
+used in AWS embedding models https://bit.ly/30kWI1f. Currently logsigmoid can be
+implemented with oneDNN by using two eltwise
+(https://oneapi-src.github.io/oneDNN/dev_guide_eltwise.html) operations
+logistic and log. Unfortunately current eltwise ops are not fusable.
+Logistic loads data from memory, make computations, stores data and next log
+loads logistic output from memory and stores data (2x loads and 2x stores). While
+it could be possibly 1x load and 1x store if these operations were combined.
+Eltwise is memory bound primitive, so additional loads and store has performance
+impact.
+
+## Proposal
+
+**Option 1 Extend exisiting eltwise with new operation:**
+
+``` cpp
+typedef enum {
+    ...
+    dnnl_eltwise_logsigmoid = number;
+    ...
+} dnnl_alg_kind_t;
+```
+
+Pros:
+- the simplest possible solution;
+- reuse existing implementation internally;
+
+Cons:
+- flexibility: extending op list each time when exisiting can be fused seems to
+be poor general solution;
+
+**Option 2 Add support for eltwise post-ops in eltwise primitive**
+
+``` cpp
+    auto eltwise_d = eltwise_forward::desc(prop_kind::forward_training,
+            algorithm::dnnl_eltwise_logistic, src_md, 0.f, 0.f);
+
+    post_ops eltwise_ops;
+    eltwise_ops.append_eltwise(scale, algorithm::eltwise_log, 0.f, 0.f);
+    primitive_attr eltwise_attr;
+    eltwise_attr.set_post_ops(eltwise_ops);
+
+    auto eltwise_pd = eltwise_forward::primitive_desc(eltwise_d, eltwise_attr, engine);
+```
+
+Pros:
+- reuse existing implementation internally;
+- flexibility - can be used each time when there will be need for fuse existing
+ops;
+
+Cons:
+- more complicated in implementation than option 1;
+
+I recommend implementing option 2, because of better flexibility.
```

Comment:
Before go with option 2, the question of general backward applicability should be explored: is it possible to apply backward computations in reversed order without spoiling the answer? If yes, how do we expect user to pass us attributes for backward: from hint or pass them explicitly? How does it work with `use_dst_for_bwd` algorithms? How does it work in case two kinds are mixed: one which supports dst on bwd, and the other one does not?

Reply:
Let me think: 
In single eltwise case for op named a, forward is defined:
d= f(s), 
so backward is 
ds = f'(s) * dd, where f'(s) is derivative, and dd input gradient computed on the preceding layer. 
If we fuse two eltwise ops named a and b (b(a)) forward is defined as
d = g(f(s)), 
so backward is 
ds = g(f(s))' * dd
General rule is:  g(f(s))' = g'(x) * f'(g(x)),
so ds =  g'(x) * f'(g(x)) * dd.
So, if your understanding of reversed order calculations in backward is that we take eltwise ops a, b, take calculated derivatives for single op a_der = f'(x), b_der =g'(x), and than apply in backward is ds = b_der(a_der(x)) *dd,  it seems that you're right and general backward applicability is not that easy. Could you please check if I didn't made any logical mistake ?

Reply:
Just a note: if the focus is on inference now, that's probably not that big of deal to skip backward (meaning that for training the integration implementation will have to split the operation in two).

On the other hand, if we want training and, say bf16, we might really want to have a alg_kind, as this will allow use to have f32 data type through out the who backward computations, while separating the operation in two will most likely lead to bf16 intermediate result. Probably worth mentioning in pros and cons of the options.

Reply:
so far I incline towards option 1 as the one that (potentially) addresses most of the issues

Reply:
I second option 1.

But I have some additional comments...

> say bf16, we might really want to have a alg_kind, as this will allow use to have f32 data type through out the whole backward computations

I'm not sure understand how fusion prevents you from having the whole backward computed in `f32`. It is would probably be challenging to do this in the current API, but eventually I think our eltwise should provide just very basic functions which could be used together to compute more complex expressions. This will also make it possible to forego all the backward and forward and just allow users to spell out the required expressions.

Reply:
@piotrchmiel I think @dzarukin touched a very important topic - depending on fuses combination, it might be very important (from perf view) to reuse intermediate computation results (done in fwd pass) to calculate bwd. In such case, how should user set up those arguments? If we assume that the user can fuse any number of eltwise ops in any order how does the mechanism of passing args should look like? Or maybe, idea is there is no intermediates reuse and some computations need to be repeated if their results are needed? This in turn may be very complicated.

Reply:
My idea is to *not* have any forward or backward operations, just `+`, `-`, `*`, `exp(x)`, ... etc. This is not proposed to be done *now*, but this is where I think it makes sense to go. This would require major overhaul of our fusion capabilities too. I'm resolving this discussion I seem to be causing more confusion than helping.